packages/gatsby-source-filesystem/src/gatsby-node.js

```diff
@@ -1,61 +1,124 @@
 const chokidar = require(`chokidar`)
 const fs = require(`fs`)
 const path = require(`path`)
-const { Machine } = require(`xstate`)
+const { Machine, interpret } = require(`xstate`)
 
 const { createFileNode } = require(`./create-file-node`)
 
 /**
  * Create a state machine to manage Chokidar's not-ready/ready states.
  */
-const createFSMachine = () =>
-  Machine({
-    key: `emitFSEvents`,
-    parallel: true,
-    strict: true,
-    states: {
-      CHOKIDAR: {
-        initial: `CHOKIDAR_NOT_READY`,
-        states: {
-          CHOKIDAR_NOT_READY: {
-            on: {
-              CHOKIDAR_READY: `CHOKIDAR_WATCHING`,
-              BOOTSTRAP_FINISHED: `CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED`,
+const createFSMachine = (
+  { actions: { createNode, deleteNode }, getNode, createNodeId, reporter },
+  pluginOptions
+) => {
+  // For every path that is reported before the 'ready' event, we throw them
+  // into a queue and then flush the queue when 'ready' event arrives.
+  // After 'ready', we handle the 'add' event without putting it into a queue.
+  let pathQueue = []
+  const flushPathQueue = () => {
+    let queue = pathQueue.slice()
+    pathQueue = null
+    return Promise.all(queue.map(createAndProcessNode))
+  }
+
+  const createAndProcessNode = path => {
+    const fileNodePromise = createFileNode(
+      path,
+      createNodeId,
+      pluginOptions
+    ).then(fileNode => {
+      createNode(fileNode)
+      return null
+    })
+    return fileNodePromise
+  }
+
+  const fsMachine = Machine(
+    {
+      id: `fs`,
+      type: `parallel`,
+      states: {
+        BOOTSTRAP: {
+          initial: `BOOTSTRAPPING`,
+          states: {
+            BOOTSTRAPPING: {
+              on: {
+                BOOTSTRAP_FINISHED: `BOOTSTRAPPED`,
+              },
+              exit: `flushPathQueue`,
             },
-          },
-          CHOKIDAR_WATCHING: {
-            on: {
-              BOOTSTRAP_FINISHED: `CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED`,
-              CHOKIDAR_READY: `CHOKIDAR_WATCHING`,
+            BOOTSTRAPPED: {
+              type: `final`,
             },
           },
-          CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED: {
-            on: {
-              CHOKIDAR_READY: `CHOKIDAR_WATCHING_BOOTSTRAP_FINISHED`,
+        },
+        CHOKIDAR: {
+          initial: `NOT_READY`,
+          states: {
+            NOT_READY: {
+              on: {
+                CHOKIDAR_READY: `READY`,
+                CHOKIDAR_ADD: { actions: `queueNodeProcessing` },
+              },
             },
+            READY: {
+              on: {
+                CHOKIDAR_ADD: { actions: `createAndProcessNode` },
+              },
+            },
+          },
+          // TODO: those two were not restricted to READY state, but maybe they should? or even maybe it should queue in NOT_READY?
```

Comment:
This is very nice illustration of why using `actions` in state machine definition is so great (opposed to what we currently have) - it made it very clear that this is not handled properly. Those definitely need to have a bit different handling in `CHOKIDAR.NOT_READY` and `CHOKIDAR.READY`

Reply:
That said, this might be not in scope of this PR (it preserves current handling) - so I'll leave this up to you if you want to attack this. If you don't - that's ok, let's just leave this TODO here

Reply:
I would prefer keeping this as is now. I can try to fix this in a followup PR though - if only you could give me hints on how the handling should differ. Should those be queued until ready?

Reply:
Yup, let's try to minimize scope of the PR. This will allow faster iteration.